packages/netlify-cms-backend-github/src/GraphQLAPI.js

```diff
@@ -0,0 +1,459 @@
+      const { is_binary, text } = data.repository.object;
+      return { is_null: false, is_binary, text };
+    } else {
+      return { is_null: true };
+    }
+  }
+
+  async retrieveContent(path, branch) {
+    const { is_null, is_binary, text } = await this.retrieveBlobObject(`${branch}:${path}`);
+    if (is_null) {
+      throw new APIError('Not Found', 404, 'GitHub');
+    } else if (!is_binary) {
+      return text;
+    } else {
+      return super.retrieveContent(path, branch);
+    }
+  }
+
+  async retrieveBlob(sha) {
+    const { is_null, is_binary, text } = await this.retrieveBlobObject(
+      sha,
+      { fetchPolicy: CACHE_FIRST }, // blob sha is derived from file content
+    );
+
+    if (is_null) {
+      throw new APIError('Not Found', 404, 'GitHub');
+    } else if (!is_binary) {
+      return text;
+    } else {
+      return super.retrieveBlob(sha);
+    }
+  }
+
+  async getStatuses(sha) {
+    const { repo_owner: owner, repo_name: name } = this;
+    const { data } = await this.query({ query: queries.statues, variables: { owner, name, sha } });
+    if (data.repository.object) {
+      const { status } = data.repository.object;
+      const { contexts } = status || { contexts: [] };
+      return contexts;
+    } else {
+      throw new APIError(`No commit found for SHA: ${sha}`, 422, 'GitHub');
+    }
+  }
+
+  async listFiles(path) {
+    const { repo_owner: owner, repo_name: name } = this;
+    const { data } = await this.query({
+      query: queries.files,
+      variables: { owner, name, expression: `${this.branch}:${path}` },
+    });
+
+    if (data.repository.object) {
+      const files = data.repository.object.entries.map(e => ({
+        ...e,
+        path: `${path}/${e.name}`,
+        download_url: `https://raw.githubusercontent.com/${this.repo}/${this.branch}/${path}/${e.name}`,
+        size: e.blob && e.blob.size,
+      }));
+      return files;
+    } else {
+      throw new APIError('Not Found', 404, 'GitHub');
+    }
+  }
+
+  async listUnpublishedBranches() {
+    console.log(
+      '%c Checking for Unpublished entries',
+      'line-height: 30px;text-align: center;font-weight: bold',
+    );
+    const { repo_owner: owner, repo_name: name } = this;
+    const { data } = await this.query({
+      query: queries.unpublishedPrBranches,
+      variables: { owner, name },
+    });
+    const { nodes } = data.repository.refs;
+    if (nodes.length > 0) {
+      const branches = [];
+      nodes.forEach(({ associatedPullRequests }) => {
+        associatedPullRequests.nodes.forEach(({ headRef }) => {
+          branches.push({ ref: `${headRef.prefix}${headRef.name}` });
+        });
+      });
+      return branches;
+    } else {
+      console.log(
+        '%c No Unpublished entries',
+        'line-height: 30px;text-align: center;font-weight: bold',
+      );
+      throw new APIError('Not Found', 404, 'GitHub');
+    }
+  }
+
+  async readUnpublishedBranchFile(contentKey) {
```

Comment:
A single query instead of multiple requests in the REST version.
The REST version will also read the entire file from the base branch just to check if it exists.